cargo/cargo_build_script.bzl

```diff
@@ -220,10 +231,19 @@ def cargo_build_script(
     cargo_build_script(
         name = "build_script",
         srcs = ["build.rs"],
-        # Data are shipped during execution.
-        data = ["src/lib.rs"],
-        # Environment variables passed during build.rs execution
-        build_script_env = {"CARGO_PKG_VERSION": "0.1.2"},
+        # Optional environment variables passed during build.rs compilation
+        rustc_env = {
+           "CARGO_PKG_VERSION": "0.1.2",
+        },
+        # Optional environment variables passed during build.rs execution.
+        # The path will be relative to the folder above bazel-out, so your
```

Comment:
I think this general change of allowing `location` expansion is a really useful one, but I don't love the interactions around path lookups.

In particular, this suggestion to traverse parent directories until you happen to find the right one. The `bazel-out` heuristic is an internal implementation detail of how bazel lays out files, and actually only happens to be true for generated files - if you did `$(execroot //some/checked/in:file)` there wouldn't be a `bazel-out`. `$(location)` and friends were designed assuming actions run in the execroot (or a runfiles tree which mirrors the execroot).

I think a reasonable approach here would be for us to change build scripts so that they executed in the execroot as their working directory (which means that these injected env vars are valid relative paths), and to recommend that crates explicitly add a `env!("CARGO_MANIFEST_DIR")` prefix to any paths they try to read at build time. We changed where build scripts run recently in https://github.com/bazelbuild/rules_rust/pull/427 so I don't feel too bad about changing it again... This approach seems fairly principled, and gives a very clear workaround for crates which face problems with this.

Any required fix pull requests may appear as slightly odd to crate authors (the author may think "the working directory _is_ `env!("CARGO_MANIFEST_DIR")` - why would I need to be explicit?"), but hopefully they'd just find it odd (and understand that alternative build tools may behave differently), rather than objectionable...

Reply:
I believe it should also work on source files, as we're only referencing "bazel-out" in the build script, which is always generated. But I take your point about this being a bit of a hack, and an approach that doesn't require hunting for execroot would be both more ergonomic and less fragile.

Perhaps one way to solve this would be to add another argument to cargo_build_script() that controls whether the working dir is execroot or the manifest dir? Crates that have bought into the Bazel ecosystem could then consume the locations directly, and third-party crates that assume cargo will set the working dir to the manifest dir for them would continue to function without patching.

Reply:
That feels like a reasonable compromise to me - we could perhaps also set an additional `BAZEL_EXEC_ROOT` env var in the manifest dir path case, to help folks who are trying to support both well to support conditional behaviour.

One challenge I've run into is crates which use [`prost-build`](https://docs.rs/prost-build), so it may be interesting to run through this as a concrete example:

With https://github.com/danburkert/prost/pull/324 (current unreleased, sadly) and this PR, you can set `build_script_env` to something like `{"PROTOC": "$(execroot @com_google_protobuf//:protoc)"}`:

In exec-root-as-working-directory mode, things should Just Work assuming all other paths (e.g. paths to the protobufs being codegen'd) are using the `$CARGO_MANIFEST_DIR` prefix properly.
In cargo-manifest-dir-working-directory mode, it would require the build script to prepend `${BAZEL_EXEC_ROOT}` to `PROTOC`, which works but is a bit annoying... Alternatively, we could extend `${pwd}`-substitution to the build script runner (which seems pretty reasonable), so that one could simply use:
`{"PROTOC": "${pwd}/$(execroot @com_google_protobuf//:protoc)"}` (or possibly `{"PROTOC": "$${pwd}/$(execroot @com_google_protobuf//:protoc)"}` - I always get lost in the escaping here!)
and the build script runner would do the right thing in both modes...

Actually, the more I talk it through, the more I'm convinced we don't need both modes, we just need `${pwd}`-substitution everywhere that we support `$(location)` substitution - how does that sound to you?

Reply:
Yeah, solving the prost-build problem was one of the motivations for this PR, as I'm using it in my project too.

Substituting pwd sounds good, I'll look into it.

Reply:
What do you think of the latest push? It adds the execroot automatically rather than relying on the user to do so - the rationale being that it's consistent with the way things like genrule() behave. But I can change it to require an explicit ${pwd} reference if you'd prefer.

And if you'd like me to squash these commits to tidy things up, please let me know.

Reply:
(I experimented with resolving execpath/${pwd} for rust_binary/rust_test(), but it didn't work - rustc_env is only available at compile time, and if we bake in the execroot, the execution stage is running in a different sandbox, and the path is no longer valid - so they will need to continue using rootpath instead)